File: python/navigation/local_planning/velocity/vo.py
"""Velocity Obstacle (Fiorini & Shiller 1998, DOI 10.1177/027836499801700706):
every tick, build one truncated cone per nearby obstacle (apex = the
obstacle's own velocity) and pick the candidate velocity closest to the
goal-seeking preferred velocity that lies outside every cone.
"""

from __future__ import annotations

from navigation.core.params import ParamSet
from navigation.core.types import Point, RobotState

from ._velocity_obstacle import DynamicObstacle, VelocityObstaclePlanner, select_sampled_velocity


class Vo(VelocityObstaclePlanner):
    def __init__(self, params: ParamSet) -> None:
        super().__init__(params)
        self._speed_samples = params.get_int("speed_samples")
        self._angle_samples = params.get_int("angle_samples")

    @property
    def name(self) -> str:
        return "vo"

    def _select_velocity(
        self,
        v_pref: Point,
        neighbors: tuple[DynamicObstacle, ...],
        statics: tuple[DynamicObstacle, ...],
        state: RobotState,
        dt: float,
    ) -> tuple[Point, tuple[tuple[float, ...], ...]]:
        x, y, _theta = state.pose
        return select_sampled_velocity(
            v_pref,
            neighbors + statics,
            (x, y),
            self._agent_radius,
            self._neighbor_dist,
            self._time_horizon,
            self._max_speed,
            self._speed_samples,
            self._angle_samples,
            apex_of=lambda o: o.velocity,
        )

File: cpp/src/local_planning/velocity/vo.cpp
#include "navigation/local_planning/velocity/vo.hpp"

#include <utility>

namespace navigation::local_planning {

Vo::Vo(core::ParamSet params)
    : VelocityObstaclePlanner(std::move(params)),
      speed_samples_(params_.get_int("speed_samples")),
      angle_samples_(params_.get_int("angle_samples")) {}

VelocitySelection Vo::select_velocity(const core::Point& v_pref,
                                      const std::vector<DynamicObstacle>& neighbors,
                                      const std::vector<DynamicObstacle>& statics,
                                      const core::RobotState& state, double /*dt*/) {
  core::Point pos{state.pose.x, state.pose.y};
  std::vector<DynamicObstacle> obstacles = neighbors;
  obstacles.insert(obstacles.end(), statics.begin(), statics.end());
  return select_sampled_velocity(v_pref, obstacles, pos, agent_radius_, neighbor_dist_,
                                 time_horizon_, max_speed_, speed_samples_, angle_samples_,
                                 [](const DynamicObstacle& o) { return o.velocity; });
}

}  // namespace navigation::local_planning
